drivers/dstate.c

```diff
@@ -1002,3 +1002,163 @@ void device_alarm_commit(const int device_number)
 		dstate_delinfo(info_name);
 	}
 }
+
+/* For devices where we do not have phase-count info (no mapping provided
+ * above), nor in the XML, we can guesstimate and report a value. This may
+ * also replace an existing value, if we've found new data, hence the bools
+ * "may_reevaluate" and the readonly flag "may_change_dstate".
+ * It is up to callers to decide if they already have data they want to keep.
+ * The "xput_prefix" is e.g. "input." or "input.bypass." or "output." with
+ * the trailing dot where applicable - we use this string verbatim below.
+ * The "inited_phaseinfo" and "num_phases" are addresses of caller's own
+ * variables to store the flag (if we have successfully inited) and the
+ * discovered amount of phases, or NULL if caller does not want to track it.
+ * Returns:
+ *   -1     Runtime/input error (non fatal, but routine was skipped)
+ *    0     Nothing changed: could not determine a value
+ *    1     A phase value was just determined (and set, if not read-only mode)
+ *    2     Nothing changed: already inited (and may_reevaluate==false)
+ *    3     Nothing changed: detected a value but it is already published
+ *          as a dstate; populated inited_phaseinfo and num_phases though
+ */
+int dstate_detect_phasecount(
+		const char *xput_prefix,
+		const int may_change_dstate,
+		int *inited_phaseinfo,
+		int *num_phases,
+		const int may_reevaluate
+) {
+	/* If caller does not want either of these back - loopback the values below */
+	int local_inited_phaseinfo = 0, local_num_phases = -1;
+	/* Temporary local value storage */
+	int old_num_phases = -1, detected_phaseinfo = 0;
+
+	if (!inited_phaseinfo)
+		inited_phaseinfo = &local_inited_phaseinfo;
+	if (!num_phases)
+		num_phases = &local_num_phases;
+	old_num_phases = *num_phases;
+
+	upsdebugx(3, "Entering set_phasecount('%s', %i, %i, %i, %i)",
+		xput_prefix, may_change_dstate, *inited_phaseinfo, *num_phases, may_reevaluate);
+
+	if (!(*inited_phaseinfo) || may_reevaluate) {
+		const char *v1,  *v2,  *v3,  *v0,
+		           *v1n, *v2n, *v3n,
+		           *v12, *v23, *v31,
+		           *c1,  *c2,  *c3,  *c0;
+		char buf[80]; /* For concatenation of "xput_prefix" with items we want to query */
+		if (!xput_prefix) {
+			upsdebugx(0, "set_phasecount(): Bad xput_prefix was passed: it is NULL - function skipped");
+			return -1;
+		}
+
+		size_t xput_prefix_len = strlen(xput_prefix);
+		if (xput_prefix_len < 1) {
+			upsdebugx(0, "set_phasecount(): Bad xput_prefix was passed: it is empty - function skipped");
+			return -1;
+		}
+
+		int bufrw_max = sizeof(buf) - xput_prefix_len;
```

Comment:
Still nitpicking on the guidelines... keep the declaration of vars at the beggining of the scope (`size_t xput_prefix_len`, `int bufrw_max` and `char *bufrw_ptr`, I'm looking at you).